File: ELECTRON/source/assets/js/memory_manager.js
const fs = require('fs');
const path = require('path');

// We’ll store each persona’s chat history in:
//  persistentdata/memory_databases/<PERSONA_NAME>/chat_history.json
function getChatHistoryPath(personaName) {
  return path.join(
    __dirname,
    '../../persistentdata/memory_databases',
    personaName,
    'chat_history.json'
  );
}

/**
 * Load chat history for a given persona.
 * Returns an array of {timestamp, role, content}.
 */
function loadChatHistory(personaName) {
  try {
    const folderPath = path.join(__dirname, '../../persistentdata/memory_databases', personaName);
    if (!fs.existsSync(folderPath)) {
      fs.mkdirSync(folderPath, { recursive: true });
    }

    const filePath = getChatHistoryPath(personaName);
    if (!fs.existsSync(filePath)) {
      // If no file yet, return empty
      return [];
    }

    const raw = fs.readFileSync(filePath, 'utf-8');
    if (!raw.trim()) {
      return [];
    }
    return JSON.parse(raw);
  } catch (err) {
    console.error(`[loadChatHistory] Error for persona "${personaName}":`, err);
    return [];
  }
}

/**
 * Save chat history array for a given persona.
 */
function saveChatHistory(personaName, historyArray) {
  try {
    const filePath = getChatHistoryPath(personaName);
    fs.writeFileSync(filePath, JSON.stringify(historyArray, null, 2), 'utf-8');
  } catch (err) {
    console.error(`[saveChatHistory] Error for persona "${personaName}":`, err);
  }
}

/**
 * Append a single new message (with role/user/assistant + content + timestamp)
 * to the persona’s chat_history.json.
 */
function appendChatMessage(personaName, role, content) {
  const history = loadChatHistory(personaName);
  const newEntry = {
    timestamp: new Date().toISOString(),
    role,
    content,
  };
  history.push(newEntry);
  saveChatHistory(personaName, history);
}

// Expose these functions for use by ipcHandlers or the Python server flow:
module.exports = {
  loadChatHistory,
  saveChatHistory,
  appendChatMessage,
};

File: ELECTRON/source/api/memory_manager.py
import os
import json
import logging
from datetime import datetime, timezone

# COMPATIBILITY NOTE: These functions are retained for compatibility with 
# the JavaScript front-end that expects chat_history.json files.
# New code should use SQLMemory directly.

def get_chat_history_path(persona_name):
    base_dir = os.path.join(
        os.path.dirname(__file__),
        "..",
        "persistentdata",
        "memory_databases",
        persona_name
    )
    if not os.path.exists(base_dir):
        os.makedirs(base_dir, exist_ok=True)
    return os.path.join(base_dir, "chat_history.json")

def load_chat_history(persona_name):
    file_path = get_chat_history_path(persona_name)
    if not os.path.exists(file_path):
        return []
    try:
        with open(file_path, "r", encoding="utf-8") as f:
            return json.load(f)
    except:
        logging.warning(f"[{persona_name}] chat_history.json corrupted. Returning empty.")
        return []

def save_chat_history(persona_name, history_array):
    file_path = get_chat_history_path(persona_name)
    with open(file_path, "w", encoding="utf-8") as f:
        json.dump(history_array, f, indent=2)

def append_chat_message(persona_name, role, content):
    history = load_chat_history(persona_name)
    entry = {
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "role": role,
        "content": content
    }
    history.append(entry)
    save_chat_history(persona_name, history)